### DISCUSSION TITLE
Search should find only matching items

### DISCUSSION THREAD

**Discussion by ildar170975** [home-assistant/frontend#12590] [Other feature requests] [answered]
Initially posted here:
https://community.home-assistant.io/t/2022-5-streamlining-settings/418477/108?u=ildar_gabdullin
https://community.home-assistant.io/t/2022-5-streamlining-settings/418477/128?u=ildar_gabdullin
https://community.home-assistant.io/t/2022-5-streamlining-settings/418477/136?u=ildar_gabdullin
https://community.home-assistant.io/t/2022-5-streamlining-settings/418477/138?u=ildar_gabdullin

Earlier (at least in 2022.4) the search feature worked as per this method:
a) the user specifies a substring -> the outputted list contains only rows with matching items (i.e. containing the substring);
b) the more precise the substring is (i.e. more letters are specified) - the less number of rows is found.

Now (2022.5) the output results may contain not only exactly matching items but similar items too.
Moreover, sometimes number of matches may INCREASE while the more precise search string is specified.
See examples in the posts above.

Probably (just my guess) these changes were added with a new search feature.

**Comment by TheHolyRoger** [comment]
@farzher it would be great if you could take a look at this please, i see you're the developer of fuzzysort :)

https://github.com/farzher/fuzzysort/issues/97

@vanackej is the one who implemented the library in this commit https://github.com/home-assistant/frontend/commit/8da73d49d7b2e160ae3b0e7b692fe9c84818741a

Demo here of the "issue" with exact matches:
![image](https://community-assets.home-assistant.io/original/4X/d/0/a/d0ad0fee488ff65941c712f5cd6549ac087b70ee.gif)

**Reply by ildar170975** [reply]
Thanks for a feedback.

Btw, (with my all respect) I do not see any need in a FUZZY search. I specify an EXACT search query - and I expect an EXACT result.
Probably, a dedicated checkbox like "Allow fuzzy search" should be provided...

**Reply by TheHolyRoger** [reply]
As I understand it, with the new search you can wrap your search term in `""` for exact matches, try it and see if it works - I'm not on 2022.5 myself yet

**Reply by ildar170975** [reply]
> with the new search you can wrap your search term in `""` for exact matches

No, it does not find any rows then.

**Reply by TheHolyRoger** [reply]
Opened a PR to fix `""` searching: https://github.com/home-assistant/frontend/pull/12591/files

**Reply by emufan** [reply]
Can only agree. if there is a or there are exakt matches (either on id or name) they are the ones to display, not any other. As in .4 and before.

**Reply by pedolsky** [reply]
With the new search scheme, the function "Add new card - by entity" has become quite redundant, which is very annoying. Before the change, I could enter _batt_ and - exclusively! - all my battery sensors were displayed. Now the output is endless and contains all kinds of entities that have absolutely nothing to do with batteries. This means that I have to go through the list one by one.

**Reply by emufan** [reply]
omg. This is really broken without additional usability given

**Reply by vanackej** [reply]
Hi. 
I originally just wanted to add exact matching but allowing multiple search terms using space separators, in order to further refine search results : 
"batt" => All entities having "batt" in it
"batt foo" => All entities having "batt" and "foo" in it, in any order
Because often when you have a lot of entities, you know some parts of the entity name/id but not always remember the exact ordering.

Then we had discussions with @zsarnett and @balloob about how the search logic should be implemented (see here: https://github.com/home-assistant/frontend/pull/10991)

In this comment (https://github.com/home-assistant/frontend/pull/10991#issuecomment-1020338266) @balloob requested that Entity Picker and Quick bar should use the same logic => So I tried to reuse the fuzzy matching for entity picker.

So the "new" entity picker is based on something very similar to previous fuzzy matcher, excepted it is not anymore based on some copy/pasted code from VSCode but on an external library that mimics Sublime text fuzzy matcher. It allows only typing the first letters of each word ("binary_sensor.garage_door_opened" => "gdo" will match), which is really a powerfull new functionnality, but clearly breaks the old exact matching.

I see some possibilities here (not exhaustive, just propositions):
- Globally revert to the old behavior, or something close to it ?
- Keep the new search for the quick bar only ? (I would personnaly be very frustrated because the old entity picker was very limited)
- Permit quotes around some search terms in order to have exact matching ?
- Just try to adress the issue that sometimes adding more caracters lead to more results (I do agree this should not happen)

We should just agree on how the search is supposed to work.

I added a lot of unit tests showing how the entity search should behave (https://github.com/home-assistant/frontend/blob/dev/test/common/string/sequence_matching.test.ts), if you can provide some comprehensive exemples of what should match and what should not I can definitly work on it.

**Reply by emufan** [reply]
I would prefer to have it as it was. exact typing. And esp. not to put in quotes or whatever to have it usable. 

If you really want to find binary_sensor.garage_door_opened by entering gdo (I would never expect to have this and see only 0,1% usecase), then introduce wildcard search. Requested here in this area in the past several times, as far as I remember/have seen:

```
g*d*o
```

**Reply by vanackej** [reply]
The behavior I originally most wanted is not this one (it is just a side effect of using the same matcher as the quick bar)
The one I reallly wanted is the first one I implemented (then removed after PR review) : "batt foo" => All entities having "batt" and "foo" in it, in any order, with exact matching for each term.

> omg. This is really broken without additional usability given

I'm sorry but I feel this is a pretty harsh statement. I can understand you don't see the usability improvement in your situation (and I agree this introduced some breaking changes and even some regressions regarding existing bahvior), but please consider the rationnals explained in the original PR and have some respect for the time and effort spent on trying to bring new features.

**Reply by TheHolyRoger** [reply]
@vanackej where's your buymeacoffee link please sir?

**Reply by vanackej** [reply]
Don't need one, I already have very good coffee at home and not looking for retribution, but thanks for asking ;-)
Sometimes trying to make good things, result is not always perfect, just expecting some kindness in comments...

**Reply by ildar170975** [reply]
@vanackej  thank you very much for all your efforts.
Do not get sad, this is a life, we all need a few steps to achieve a perfect output.

In short - it would be great to have 2 possibilities:
- exact match search
- fuzzy search

This may be achieved:
- either by using a "enable fuzzy search" checkbox (located somewhere close to the input field)
- or by using a syntax

1st way is clear: checkbox OFF -> search only exact matches (as it was recently), checkbox ON -> use your smart algorithms.
Probably the **default** checkbox's state should be always OFF.

2nd way:
- you can use "zzzzz.xxxxx_1234_yyyyy" - i.e. quotes - to specify an exact search;
- otherwise a fuzzy search is used.

As for me, I vote for the 1st way....
Sometimes we need a simple and explicit method of doing something - then I switch off unneeded experimental features & AI.

**Reply by emufan** [reply]
> pretty harsh statement.
Sorry, if you get it wrong. Don't expect, that not everyone is a native speaker here. I'm always open and of course everything I write is my personal feelings. If I like something, that I write it and if I think that it is broken (and for me it is), I write it as well. Don't search for harshness, if this was not meant that way. 💙

**Comment by vanackej** [comment]
@zsarnett Feedbacks on the new entity picker behavior seems globally negative.
I can, if you are ok with it:
- Remove fuzzy search from entity picker/filtering
- Just reintroduce the simple behavior I originally wanted to implement : If search terms are separated with spaces, ensure all terms are included in the id or name of the entity (which I think is a real improvement over the simple "contains" initial approach, with no breaking change, just a slight improvement)
- Optionnaly, add some simple wildcard mecanism like suggested by @emufan : nice to have, easy to implement, but not really mandatory
Let me know what you think should be done, I'd be happy to propose a PR quickly.

**Reply by bschatzow** [reply]
What gets lost in the development is that different sections have different target users. People in general using developers tools, understand what they are looking for. Finding more options does not help ( answer is lost in the forest). Finding complete information is what is needed, I. E. Entity id etc. Maybe have two search boxes will answer to both groups?

**Reply by ildar170975** [reply]
Alternatively as I wrote before - use a "Fuzzy search" checkbox.
But then this checkbox may have different default values:
- OFF - in the Dev tab (and some other places probably)
- ON - in ?????? tab (cannot imagine where it should be default)

**Comment by vanackej** [comment]
I don't know what to think, so many options/opinions here :-)

My simple initial intention, as an end user, is to make the entity search a little more functionnal for this simple (and I think very frequent) use case:
- I want to find one specific entity, let's say a temperature sensor located in roomX
- I have a lot of temperature sensors (so typing "temp" in search return too many results)
- I have a lot of entities in room X (so typing "roomX" in search also return too many results)
- I don't know in what order/the exact naming is, so I'm not able to find easily with the previous "includes" filtering
 
=> I just want to type "temp" to filter temperature sensors, and then "roomX" to filter on the location

Another example : I want one of my hvac units total consumption entity. Just want to type, in any order, the words "consumption roomName total" to get it. 
The currently (pre 2022.5) entity filtering totally unable to achieve this kind of search (which are the ones I always need)

I'm not sure having advanced search mode/switching between search mode with checkboxes would be really that usefull, and it would add a lot of complexity in the user interface and underlying code

**Reply by ildar170975** [reply]
I totally agree with this case. Fuzzy search may be useful, let's not just kill it.

Then what about this algorithm:
1) xxx_xxx.xxx_1234_xxx - use exact search
2) xxx_xxx SPACE xxx_1234_xxx - use fuzzy search


So, if SPACE is defined at least one time - use fuzzy.

**Reply by TheHolyRoger** [reply]
@vanackej how about using exact match first, sticking those results at the top, then fuzzysort matches after that - which will also mean that when there's no exact match you'll always get fuzzy with the handy typo checking etc

**Reply by vanackej** [reply]
I'm not so sure it would be such a good idea switching from exact to fuzzy search using space.
I think one global requirement for a search fonctionnality is that adding more characters/terms to a search box should always reduce the number of displayed results, not potentially add new results. 
Or else it is really confusing for the end user (exactly like illustrated in @TheHolyRoger screen record here https://github.com/home-assistant/frontend/discussions/12590#discussioncomment-2690851)

**Reply by vanackej** [reply]
> @vanackej how about using exact match first, sticking those results at the top, then fuzzysort matches after that - which will also mean that when there's no exact match you'll always get fuzzy with the handy typo checking etc

That's exactly how the new search scoring works : exact matches have better score and should be at the top of the results.
the problem here is that results are not ordered by score in those views, but by name or whatever sort criteria is defined on the result table, independently for search score.
So results with low scores (= less relevant) can appear before best matches

![image](https://user-images.githubusercontent.com/534429/166912288-10af6d64-468d-4a49-aa89-1e8856a32ac5.png)

In quick bar, results are ordered by score, so there is almost no issue here because most relevant results appear first

**Reply by ildar170975** [reply]
> Or else it is really confusing for the end user (exactly like illustrated

Yes, that was my video, very confusing. And gives different results for "add more chars" & "remove added chars".



> how about using exact match first, sticking those results at the top

Not good...
Sometimes we (at least me) use a `or.xxxx_123_` search mask to see **how many and which exactly** entities I have for this mask.
So, excess rows may confuse.

**Reply by TheHolyRoger** [reply]
> > @vanackej how about using exact match first, sticking those results at the top, then fuzzysort matches after that - which will also mean that when there's no exact match you'll always get fuzzy with the handy typo checking etc
> 
> That's exactly how the new search scoring works : exact matches have better score and should be at the top of the results. the problem here is that results are not ordered by score in those views, but by name or whatever sort criteria is defined on the result table, independently for search score. So results with low scores (= less relevant) can appear before best matches
> 
> ![image](https://user-images.githubusercontent.com/534429/166912288-10af6d64-468d-4a49-aa89-1e8856a32ac5.png)
> 
> In quick bar, results are ordered by score, so there is almost no issue here because most relevant results appear first

That doesn't seem to be what's going on here though, this view isn't sorted except by the score is it?

![image](https://community-assets.home-assistant.io/original/4X/5/3/e/53e60445ddd25d1e64e8262f2d754eb2c1741908.png)

**Reply by vanackej** [reply]
> And gives different results for "add more chars" & "remove added chars".
> 
> >

Yes, I saw it too, really strange. I don't think it's related to the search algorithm by itself, more a kind of frontend bug. I can have a look at it too

> Sometimes we (at least me) use a `or.xxxx_123_` search mask to see **how many and which exactly** entities I have for this mask.
> So, excess rows may confuse.

I really think the simpler option for now is the one I proposed [here](https://github.com/home-assistant/frontend/discussions/12590#discussioncomment-2692531) and initially in the first version of my PR. It is a really conservative option, that restore previous behavior and just allows the use cases I mentionned [here](https://github.com/home-assistant/frontend/discussions/12590#discussioncomment-2692947)

The fuzzy search method would still be somewhere used in the codebase (in the Quickbar only for now), and could maybe be reintroduced later in other parts of the frontend after having a better view on the topic

**Reply by vanackej** [reply]
> That doesn't seem to be what's going on here though, this view isn't sorted except by the score is it?

Yes, this view is ordered by score. I think the underlying library is using the dot separator as a special char, not sure about it.
I can obviously have a look to make the ordering more coherent, as it would be logical that the third entry should be at the top of the results. I will reproduce the case with an unit test a see how to improve the scoring (whatever decision is made on the more general issue : should we globally use fuzzy matching or not on entity search)

**Reply by ildar170975** [reply]
> I really think the simpler option for now is the one I proposed [here](https://github.com/home-assistant/frontend/discussions/12590#discussioncomment-2692531)

Agree with this. Seems to be the same what I wanted [here](https://github.com/home-assistant/frontend/discussions/12590#discussioncomment-2692970).



> fuzzy search method would still be somewhere used

As I understood, the fuzzy method will be used when a SPACE char is defined? If so, this method will be successfully used.

**Reply by TheHolyRoger** [reply]
> > That doesn't seem to be what's going on here though, this view isn't sorted except by the score is it?
> 
> Yes, this view is ordered by score. I think the underlying library is using the dot separator as a special char, not sure about it. I can obviously have a look to make the ordering more coherent, as it would be logical that the third entry should be at the top of the results. I will reproduce the case with an unit test a see how to improve the scoring (whatever decision is made on the more general issue : should we globally use fuzzy matching or not on entity search)

For your next trick I'd like you to make it so I can search entities by imagining them in my head please sir

**Reply by bschatzow** [reply]
I always found search with wild cards easy to use.  I.e *123 finds all with 123 at the end *123* finds all with 123 anywhere in it, 123* must begin with 123.  Using the ? for a single character different (1?3) combined with the * has always worked for me.  This search has  been around for many many years and it seems easy to understand for most users.

**Reply by farzher** [reply]
most of the search issues are from frontend bugs or sorting alphabetically, not the algo.
but `er.life360` returning bad results is a fuzzysort issue. searching starting from the middle/end of a word isn't a supported use-case.
er really doesn't want to match to tracker. while `tr.life360` would match perfectly.

@vanackej @TheHolyRoger

**Reply by ildar170975** [reply]
> er really doesn't want to match to tracker. while `tr.life360` would match perfectly.

`er.life360` is a substring of `device_tracker.life360_xxxxx`.

**Reply by farzher** [reply]
> > er really doesn't want to match to tracker. while `tr.life360` would match perfectly.
> 
> `er.life360` is a substring of `device_tracker.life360_xxxxx`.

yup. i'll definitely be fixing this. the first change to the algorithm in 5 years. but don't expect that anytime soon, and feel free to replace fuzzysort; and don't ask me how nobody's complained about this over the past 5 years. ¯\\_(ツ)_/¯

**Reply by ildar170975** [reply]
> don't ask me how nobody's complained about this over the past 5 years

I have was been using search in the Entities list, "Set state" panel, the "History" page for last 2 months, it worked perfectly))).

**Reply by farzher** [reply]
> > don't ask me how nobody's complained about this over the past 5 years
> 
> Mmmm, I have was been using this search for last 2 months, it worked perfectly)))

oh sorry, i'm the dev of https://github.com/farzher/fuzzysort. this project just started using my search. i don't work here :)

**Reply by vanackej** [reply]
@farzher Thanks for having a look and the explaination about how the fuzzysort is intended to works

It confirms what I see in my tests :

```
Search for er.life360_iphone_6 in binary_sensor.battery_charging_life360_iphone_6,: -51028
Search for er.life360_iphone_6 in sensor.battery_life360_iphone_6,: -21012
Search for er.life360_iphone_6 in device_tracker.life360_iphone_6,: -23012

Search for tr.life360_iphone_6 in binary_sensor.battery_charging_life360_iphone_6,: No score
Search for tr.life360_iphone_6 in sensor.battery_life360_iphone_6,: No score
Search for tr.life360_iphone_6 in device_tracker.life360_iphone_6,: -33
```

When matching sequence is at the end of the word, thats not really that it is not "supported", but score is a lot lower (then it depends on the minimum accepted score to be considered as a match) => Results appears to be out of the logical order

A simple workaround (until @farzher maybe fix this in fuzzysort library) is to first test if the search term is included in the result before using fuzzysort :
```
Search for er.life360_iphone_6 in binary_sensor.battery_charging_life360_iphone_6,: -51028
Search for er.life360_iphone_6 in sensor.battery_life360_iphone_6,: -21012
Search for er.life360_iphone_6 in device_tracker.life360_iphone_6,: 0
```

**Comment by zsarnett** [comment, marked as answer]
Hi @farzher 

Thanks for the insight into the Algorithm from FuzzySort! I do think we can continue to utilize this algorithm with a few tweaks to some of the scorings. This is possible using your currently extendable options. 

I will discuss with @vanackej who created the original PRs for this and find a better method. I do think if the filter string is included in one of the strings in the list then we should return that to the top of the list.

So we can do something where we remove those from the fuzzy sort options, before filtering them using our own method or possibly updating the scoring option. I haven't fully dived into this yet but we will make the change for the next release for sure

Thanks to everyone who has been discussing this and giving feedback!

**Reply by vanackej** [reply]
I actually already implemented this strategy and added tests for it.
It seems to fix to some of the issues here, but not all expected usages are supported (i.e. it does not fully restore previous simple behavior in some situations).

**Reply by zsarnett** [reply]
Restoring the previous behavior is not the goal. We should just make sure it provides filtering of entities in a way that is predictable for most users.

**Reply by bschatzow** [reply]
Is predictable the goal or should it be usable for the intent?  

I feel anything in developer's tools should always give more information than the typical user wants ( or thinks they want). When I am in developer's tools  ( for HA or for any other application for that matter) I am trying to solve an issue that I am having. This issue came about from trying to search for an entity and the ID is missing and than the search was not specific enough to find what the user was looking for.

**Reply by zsarnett** [reply]
We are going to revert these changes in a patch release tomorrow.

We can reevaluate next release

**Reply by ildar170975** [reply]
> in a patch release tomorrow.

Not in HA 2022.5.2 , I guess.

**Reply by zsarnett** [reply]
In a patch release tomorrow 

Edit: Released in the .3 Patch

**Reply by vanackej** [reply]
@zsarnett do you want me to restore previous behavior (but keeping infrastructure change in order to allow latter improvements ?). Tell me if I can help on this

**Reply by zsarnett** [reply]
@vanackej if you want to try the fuzzy sort library again then we can see what we can customize about it

Or we can just update the current implementation from vscode and add in functionality

**Reply by bschatzow** [reply]
I do not see it fixed in 5.3. I just tried the drop down in "people" typed in gps, should have found 4 sensors, found none.

**Reply by zsarnett** [reply]
It's reverted for me in the .3 release. Options not showing is more of a case for the revert state than the non reverted state.

**Comment by MRobi1** [comment]
Just jumping in as I had opened another issue on this same topic here https://github.com/home-assistant/core/issues/71435

I hope this can be fixed in an upcoming release because the search function as it is is barely usable.

**Reply by sonnstein** [reply]
absolutely agree! It is of no worth to have a search result, with 50% result, that has virtually no (some mystery) relation to the search string!
